subsys/spm/secure_services.c

```diff
@@ -47,6 +53,34 @@ int spm_secure_services_init(void)
 }
 
 
+#ifdef CONFIG_SPM_SERVICE_READ
+__TZ_NONSECURE_ENTRY_FUNC
+int spm_request_read(void *destination, const void *source, size_t num)
+{
+	static const u32_t read_ranges[] = {
+#ifdef PM_MCUBOOT_ADDRESS
+		/* Allow reads of mcuboot metadata */
+		PM_MCUBOOT_PAD_ADDRESS,
+		PM_MCUBOOT_PAD_ADDRESS + PM_MCUBOOT_PAD_SIZE
+#endif
+	};
+
+	if (destination == NULL || num <= 0) {
+		return -EINVAL;
+	}
+
+	for (int i = 0; i < sizeof(read_ranges); i += 2) {
+		if ((u32_t)source > read_ranges[i] &&
+		    (u32_t)source < read_ranges[i+1]) {
+			memcpy(destination, source, num);
```

Comment:
🐛 You must check that `source` and `source + len` are within a range, otherwise the caller is able to read the whole RAM by passing any source in a range :)

Reply:
Good catch 🥇  See proposed fix